src/store/plugins/load-geojson-style-and-data.plugin.js

```diff
@@ -63,29 +64,62 @@ function clearAutoReload(layerId) {
 
 /**
  * @param {GeoAdminGeoJsonLayer} geoJsonLayer
- * @returns {Promise<GeoAdminGeoJsonLayer>}
+ * @returns {{ controllers: [AbortController]; clone: Promise<GeoAdminGeoJsonLayer> }}
  */
-async function loadDataAndStyle(geoJsonLayer) {
+function loadDataAndStyle(geoJsonLayer) {
     log.debug(`Loading data/style for added GeoJSON layer`, geoJsonLayer)
-    const clone = geoJsonLayer.clone()
-    try {
-        const [{ data: style }, { data }] = await Promise.all([
-            load(geoJsonLayer.styleUrl),
-            load(geoJsonLayer.geoJsonUrl),
-        ])
-
-        // as the layer comes from the store (99.9% chances), we copy it before altering it
-        // (otherwise, Vuex raises an error)
-        clone.geoJsonData = data
-        clone.geoJsonStyle = style
-        clone.isLoading = false
-        return clone
-    } catch (error) {
-        log.error(`Error while fetching GeoJSON data/style for layer ${geoJsonLayer?.id}`, error)
-        clone.isLoading = false
-        clone.errorKey = 'loading_error_network_failure'
-        clone.hasError = true
-        return clone
+
+    const style = load(geoJsonLayer.styleUrl)
+    const data = load(geoJsonLayer.geoJsonUrl)
+
+    return {
+        controllers: [style.controller, data.controller],
+        clone: Promise.all([style.response, data.response])
+            .then(([{ data: style }, { data }]) => {
+                const clone = geoJsonLayer.clone()
+                // as the layer comes from the store (99.9% chances), we copy it before altering it
+                // (otherwise, Vuex raises an error)
+                clone.geoJsonData = data
+                clone.geoJsonStyle = style
+                clone.isLoading = false
+                return clone
+            })
+            .catch((error) => {
+                log.error(
+                    `Error while fetching GeoJSON data/style for layer ${geoJsonLayer?.id}`,
+                    error
+                )
+                const clone = geoJsonLayer.clone()
+                clone.isLoading = false
+                clone.errorKey = 'loading_error_network_failure'
+                clone.hasError = true
+                return clone
+            }),
+    }
+}
+
+let pendingPreviewLayer = null
+
+async function loadAndUpdatePreviewLayer(store, layer) {
+    cancelLoadPreviewLayer()
```

Comment:
I'm wondering if we shouldn't check which layer we are loading. Because if we are hover a GeoJSON and then clicking it very quickly, this means this will cancel the initial load of data/style for the layer, and we will fire it once more to show it definitively on the map...
I don't know if this would bring too much complexity here, your thoughts?

Reply:
@pakb good point, however this would add much complexity. Currently the preview layer object and the active layer object are two different object, therefore when the layer is added it reload the geojson data and style for the new object. We could in theory reuse the preview object instead of creating a new object, but I think this would complexify the code and could bring more trouble. If the geojson style and data have proper cache-control header (I check a few meteoswiss layers) then we don't have any performance issue as the second load is done by the browser cache.